File: js/src/surrogate.js
// UTF-16 surrogate pairs — code points above the Basic Multilingual Plane.
//
// Failure mode #8 (utf16-surrogate-split): a JavaScript string is a sequence of
// UTF-16 code units, and any code point above U+FFFF is stored as two units (a
// surrogate pair). Rare CJK kanji live up there — 𠮷 (U+20BB7, the official 吉
// in 𠮷野家 / Yoshinoya), 𩸽 (ほっけ), and the CJK Extension B block — as do all
// emoji. So `"𠮷".length` is 2, and `.slice`, `.substring`, `charAt`, and `s[i]`
// all index by code unit. Cut at an odd boundary, or count `length` as a
// character limit, and you split the pair into a lone surrogate: a broken glyph,
// a U+FFFD on round-trip through UTF-8, a database value truncated to garbage.
//
// This is distinct from the byte-boundary crash (#2): that one is about UTF-8
// byte offsets crossing into a multi-byte sequence; this one is about UTF-16
// code-unit offsets crossing into a surrogate pair. Slicing by code point fixes
// both.

/** Buggy reference: slice by UTF-16 code unit, which can split a surrogate pair. */
export function utf16Slice(str, nUnits) {
  return str.slice(0, nUnits);
}

/** Correct: slice by code point so an astral character stays whole. */
export function codePointSlice(str, nChars) {
  return Array.from(str).slice(0, nChars).join("");
}

/** UTF-16 code-unit count — an astral character counts as 2. */
export function utf16Length(str) {
  return str.length;
}

/** Code-point count — what a user would call the number of characters. */
export function codePointLength(str) {
  return Array.from(str).length;
}

/** True if the string contains no unpaired surrogate (well-formed UTF-16). */
export function isWellFormed(str) {
  for (let i = 0; i < str.length; i++) {
    const u = str.charCodeAt(i);
    if (u >= 0xd800 && u <= 0xdbff) {
      // high surrogate: must be followed by a low surrogate
      const next = i + 1 < str.length ? str.charCodeAt(i + 1) : 0;
      if (next < 0xdc00 || next > 0xdfff) return false;
      i++;
    } else if (u >= 0xdc00 && u <= 0xdfff) {
      // low surrogate with no preceding high surrogate
      return false;
    }
  }
  return true;
}

File: go/surrogate.go
package cjkfixtures

import (
	"strings"
	"unicode/utf16"
)

// UTF-16 surrogate pairs (failure mode #8, utf16-surrogate-split). A Go string
// is UTF-8, so this hazard appears when the string crosses into a UTF-16 host: a
// Windows API (syscall.UTF16FromString), a JavaScript or Java bridge, a SQL
// Server NVARCHAR column, or a JSON \uXXXX escape. There, every code point above
// U+FFFF — rare kanji like 𠮷 (U+20BB7, the kanji in 𠮷野家) and all emoji — is
// two UTF-16 code units (a surrogate pair). Counting or slicing by code unit
// splits the pair into a lone surrogate, which decodes back to U+FFFD. Slicing
// by rune keeps the character whole. This mirrors JavaScript, where .length and
// .slice index by UTF-16 unit natively.
//
// It is distinct from the byte-boundary crash (#2): that one crosses a UTF-8
// byte boundary; this one crosses a UTF-16 code-unit boundary.

// Utf16UnitLength is the number of UTF-16 code units; an astral rune counts as 2.
func Utf16UnitLength(s string) int {
	return len(utf16.Encode([]rune(s)))
}

// RuneLength is the number of code points.
func RuneLength(s string) int {
	return len([]rune(s))
}

// Utf16UnitSlice is the buggy reference: slice by UTF-16 code unit, which can
// split a surrogate pair. A lone surrogate decodes to U+FFFD.
func Utf16UnitSlice(s string, nUnits int) string {
	u := utf16.Encode([]rune(s))
	if nUnits > len(u) {
		nUnits = len(u)
	}
	return string(utf16.Decode(u[:nUnits]))
}

// RuneSlice is correct: slice by rune so an astral character stays whole.
func RuneSlice(s string, nRunes int) string {
	r := []rune(s)
	if nRunes > len(r) {
		nRunes = len(r)
	}
	return string(r[:nRunes])
}

// HasReplacement reports whether the string contains U+FFFD, a sign a surrogate
// pair (or a byte sequence) was split.
func HasReplacement(s string) bool {
	return strings.ContainsRune(s, 0xFFFD)
}
